### ISSUE TITLE
Success Criterion 2.4.7 - Focus Visible - Level AA

### ISSUE THREAD

**Issue by JJdeGroot** [w3c/matf#51]
WCAG2ICT guidance: https://www.w3.org/TR/wcag2ict-22/#focus-visible

Share your thoughts for applying to mobile apps as a comment below.

**Comment by JJdeGroot** [comment]
**Discussed in today's meeting.**

<details>
<summary>Minutes for 28 August 2024</summary>

[Source](https://www.w3.org/2024/08/28-matf-minutes.html#t04)

We need to define "keyboard operable user interface"

<Jamie> was there a reasin to need large variation from WCAG?

ACTION: Add minutes from variations to discussed SC's

<julianmka> Techniques for keyboard focus in Android ViewUI: https://github.com/cvs-health/android-view-accessibility-techniques/blob/0ffd2aad8f8eada9893295ac47cc9ee9bf21372d/doc/dynamicbehaviors/CustomFocusIndicators.md#L4

<julianmka> Techniques for keyboard focus in Android Compose: https://github.com/cvs-health/android-compose-accessibility-techniques/blob/main/doc/interactions/CustomFocusIndicators.md

Jamie quintinb it's not really our job to tell developers

<julianmka> ^^ I don't have the depth to evaluate these techniques but interested in others' takes

Jamie This is one of the few SC's where an external keyboard is required.
</details>

**Comment by jamieherrera** [comment]
In reviewing this again if I recall, earlier we were tripping up on "keyboard operable... interface" as part of a leftover discussion from other SC about keyboards, where we want to allow for other AT to be substituted for external keyboard. Here, though, if we add a note that this is intended to be an SC specifically about the external keyboard use (should we specify when NOT also using a screen reader?) then we could probably leave the normative text as-is. 

Otherwise the notes from last time also asked why we made this large variation... maybe it was a typo?

**Comment by JJdeGroot** [comment]
Briefly mentioned in today's meeting, will be back on agenda next week.

**Comment by pauljadam** [comment]
Focus Visible is tested using an external keyboard and ensuring all the UI controls have visible keyboard focus when focused with the keyboard.

**Comment by tgravemaker** [comment]
I agree with Jamie that can leave the text as-is with a note about sole use of a keyboard for this SC (instead of the broader group of assistive technologies or a combination of them). I do however think that this SC could benefit from some extra clarification via the Understanding Docs. Especially since developers are limited in the implementation of those borders it would be helpful to determine when this SC passes or fails.

**Comment by JJdeGroot** [comment]
Discussed in today’s meeting.

<details>
<summary>26 February 2025</summary>

[Source](https://www.w3.org/2025/02/26-matf-minutes.html#17f7)

~~~text
JJ Focus visble - JJ confused about keyboard operable user interface. This seems to focus on keyboard as opposed to keyboard interface

This SC seems to be written for keyboard and not keyboard interface

@JJ reading comments on GitHub

<pauljadam> Visible does not mean it has a certain contrast, that's not defined.

@JJ but customisable - then when developers have the capacity to change the focus indicator, it should be visible. But JJ means the system indicator

pauljadam iOS won't let you chnage the indicator

@JJ it's not possible to modify the system

<Jamie> qw

Joe_Humbert - we have said this only applies to keyboard, but switch access also uses the keyboard interface. This might not be the same on iOS and Android

by the way, Switch Control / Access is using the KB interface

I've been doing this with Ally Keys

Joe_Humbert then why did FKA preceed them?

pauljadam probably due to numbers (sorry for paraphrasing)

@JJ we need to be careful about wording

pauljadam on switch control - if it's keyboard accessible, switches work fine. I wouldn't change the definition. All you need to test is keyboard. @JJ confirms

Jamie have we confirmed that?

Switches allow for actions, which Keyboard does not, but other than that I haven't noticed much

Focus and selection are the same

pauljadam are we extending the scope of wcag if we explicitly require switch access?

Jamieyes

JJ mobile is changing a lot of the landscape

pauljadam WCAG doesn't say to test for focus color in switch control

@JJ I wonder how we should be reading this sc

<JJ> "Any" keyboard operable user interface has "a mode" of operation where the keyboard focus indicator is visible.

Detlev since switch is next / activate but keyboard allows for more granular control. Does that map equivalently?

@JJ are we just assuming switch access works - it's confused me in the past. 2.1.1 doesn't mean that switch and touch has to work the same. If the keyboard has equivalent control

pauljadam you could even use the enter key instead of a submit button

ACTION: Research on Android + iOS the behavior of Switch Control and External Keyboard focus indicators

quintinb the Switch interface allows for "build your own keyboard" as well as auto select etc.

<Joe_Humbert> 1+ to quintinb about switch access is very customizable

<pauljadam> That's why you keep it to Keyboard

<pauljadam> If you add accessibility Actions on iOS it works for VoiceOver, Keyboard, Switch all the same

I learned what I know from pauljadam's stuff, lol

<pauljadam> :)

<pauljadam> keep as is will be my response mostly :)

<quintinb> +1 pauljadam

JJ we could add a note about best practice regarding accessibility interface

<pauljadam> we going to start talking about Eye Tracking too? probably too hard to discuss every possible AT on Mobile

<JJ> Keep the text as is, but add note about best-practices for focus visible for assistive technologies?

<pauljadam> So regular WCAG does not say "Make sure you do switch controls too" But Mobile WCAG needs to say that?

julianmka we should keep to a similar approach that we've been taking regarding mobile being different. We should be aware of what the system supports and keep with intent rather than being explicit on every possible AT

julianmka windows doesn't support switch control

<JJ> Poll: Keep text as is, add note about supporting focus indicators for other types of assistive technologies?

<Tim> +1

<GleidsonRamos> +1

<pauljadam> Keep as is, no note needed

<Jamie> +1 plus note

<Tanya> +1

<Detlev> +1

<quintinb> +1

<Illai> +1

<julianmka> +1

<Carol> +1

<pauljadam> What if Apple or Android changes later?

<pauljadam> Are we going to define everything possible for a developer?

Joe_Humbert do we need to acknowledge that devs can change the original focus indication

JJ we could use a similar note to 1.1.1 where not every platform does something

<pauljadam> I'm not sure it's worth the effort of if we'd have full accuracy if we start defining what is possible and what is not possible for each platform.

ACTION: Add second note: Not all mobile platforms provide a way to programmatically change the keyboard focus indicator.
~~~
</details>

- ACTION: Research on Android + iOS the behavior of Switch Control and External Keyboard focus indicators
- ACTION: Add second note: Not all mobile platforms provide a way to programmatically change the keyboard focus indicator.

**Comment by julianmka** [comment]
# Focus style for keyboard vs switch - Android

Device info: Samsung zFlip 6, Android 14, OneUI 6.1.1

Summary: Switch control focus style and keyboard focus style are independent. Changing switch focus style did not affect keyboard focus.

Image 1: keyboard navigation focus shown in settings app. Focused element darkens and there is no outline. 
![Screenshot_20250226_092934_Gallery.png](https://github.com/user-attachments/assets/61ede5c2-ac93-4172-bc26-498f5e7d0dcf)

Image 2: keyboard navigation focus in the app library/home screen. The focused element has a thin, light blue outline. Switch control focus style is set to a red outline. (And the switch control outline style is considerably thicker and uses both a light and a dark color.)
![Screenshot_20250226_093949_One UI Home.png](https://github.com/user-attachments/assets/68482860-2ef4-489b-a00e-d33c7794810c)

Image 3: switch control focus shown in settings app. Focused element is surrounded by a red highlight. 
![Screenshot_20250226_093017_Accessibility.png](https://github.com/user-attachments/assets/860c66d0-f0e0-482f-8c5e-dd8539eb730f)

**Comment by pauljadam** [comment]
Are we trying to research the assumption that if you make something keyboard operable it will also be switch control operable?

**Comment by julianmka** [comment]
@pauljadam My comment was in response to the action from today's meeting, "ACTION: Research on Android + iOS the behavior of Switch Control and External Keyboard focus indicators." There was discussion about whether the focus styles for switch control and keyboard navigation were linked.

**Comment by JJdeGroot** [comment]
Thanks @julianmka ! And to @pauljadam  - yes, that, plus have a reference of focus indicators on mobile platforms

**Comment by pauljadam** [comment]
@julianmka Keyboard Focus Color Options on iOS:

![Image](https://github.com/user-attachments/assets/fbcb4275-fef9-4441-a716-da306e11d781)

Keyboard users can choose Default, Gray, White, Blue, Red, Green, Yellow, Orange.

But if you select High Contrast then the focus outline will always be Black. 

![Image](https://github.com/user-attachments/assets/b9245ec4-e700-40ca-bd42-8116e61b80ce)

Switch Control Cursor Color Options on iOS:

![Image](https://github.com/user-attachments/assets/9bdec03b-e81d-4831-ac36-4f21f5af60cb)


Switch control users can choose Blue, Red, Green, Yellow, Orange.

**Comment by detlevhfischer** [comment]
@pauljadam wrote
> Are we trying to research the assumption that if you make something keyboard operable it will also be switch control operable?

I think it would be useful to verify that asserting keyboard accessibility also implies content is switch-accessible for the reasons I mentioned in the call: In many cases full keyboard operability can only be asserted if you include at a minimum both TAB / SHIFT TAB, arrow keys (all four) and TAB / SHIFT + TAB with CTRL modifier. Given that the normative text of 2.1.1 does not say anything about **what** keys to use to conform to 2.1.1, the assumption must be that if I can use some key or key combination to get to any and all interactive elements in the view, that would constitute a minimal pass. 

Even given customisation options for switch users which some users may need and use, I assume the common way of using switch is more constrained: 

1. either a two button mode where one button advances the focus and the other activates, 
2. or a mode where you have a timed sequential movement through interactive elements and activate once you see the element you want has received focus. 

So in those base scenarios which I assume would need to be supported, there are no modifier keys or different keys (like TAB and arrow keys). I will try to run some tests on app examples where keyboard access needs these different operations, including modifier keys.

**Comment by pauljadam** [comment]
I've not been able to figure out how to make something that is accessible to full keyboard access but not accessible to switch control. 

This screenshot shows a bad functional image example on iOS that is not coded as a button and instead uses an Image with a Tap gesture and it is accessible to both full keyboard access and switch control.
![Image](https://github.com/user-attachments/assets/bf674b2d-c2ac-4a32-a602-eda60fe56bb4)


Full keyboard access is also able to focus on the image with a tap gesture and tap it using the keyboard alone. 

![Image](https://github.com/user-attachments/assets/1b00320f-29fc-4f86-bc4b-83cb3e283229)

**Comment by maryjom** [comment]
This isn't about what the focus cursor looks like, but about whether or not it is visible when an interactive component receives keyboard focus. In WCAG, this was trying to solve the problem where the authored content was turning off the keyboard focus indicator or making it match the background so it wouldn't be visible to the user. IMO, the question to answer here is whether that type of capability exists in the mobile space.

**Comment by JJdeGroot** [comment]
Thanks for the clarification @maryjom 
At this moment, there is no way to turn off the keyboard focus indicator on Android and iOS. Developers are only able to add additional indicators. Also, the styling of the default keyboard focus indicator (and other assistive tech) cannot be adjusted.

Maybe you could hide the indicator by changing the `accessibilityFrame`. E.g. if that's zero `(0, 0, 0, 0)` perhaps no indicator will be drawn?

Is that something you can verify on iOS, @pauljadam ?

**Comment by jha11y** [comment]
<details>
<summary>Notes from June 11 MATF meeting</summary>
<QuintinB> We need to be clear that platforms may provide certain visible focus elements via the platform itself, but we need to be cautious re: x platform that we're no so lenient that that becomes an excuse
</details>
Will be discussed further next meeting
